### PR TITLE
arm64: Initial implementation of CONFIG_BUILD_KERNEL

### PR DESCRIPTION
**Pull request by pussuw**
## Summary
arm64: Initial implementation of CONFIG_BUILD_KERNEL
    
This is the initial version for kernel mode build on the arm64 platform.
It works much in the same way as the risc-v implementation so any
highlights can be read from there.
    
Features that have been tested working:
- Creating address environments
- Loading init (nsh) from elf file
- Booting to nsh
- Starting other processes from nsh
- ostest runs to completion
    
Features that are not tested / do not work:
- SHM / shared memory support
- Kernel memory mapping (MM_KMAP)
- fork/vfork
    
An example qemu target is provided as a separate patch:
tools/configure.sh qemu-armv8a:knsh
## Impact
Add kernel mode for arm64, no impact on flat mode operation 
## Testing
qemu-armv8a:knsh
imx93 custom target (on real hardware)

### CODE REVIEW

**Review by jerpelea — APPROVED** [review]

**Comment by acassis** [comment]
Hi @xiaoxiang781216 seems like CI is frozen! 2h and it didn't even pass the initial 4 checks:

image

**Comment by pussuw** [comment]
`./cxxtest_main.cxx:31:10: fatal error: fstream: No such file or directory`
I think this is due to a missing C++ library, not sure why my build passes locally but CI fails. I think I'll just disable cxxtest.

```
Error: fpu.c:55:8: error: #warning "FPU test not built; Only available in the flat build (CONFIG_BUILD_FLAT)" [-Werror=cpp]
   55 | #      warning "FPU test not built; Only available in the flat build (CONFIG_BUILD_FLAT)"
      |        ^~~~~~~
cc1: all warnings being treated as errors
```
This I'm not sure how to fix; ostest builds the FPU test for BUILD_KERNEL and BUILD_PROTECTED and then gives this warning about how the test does not support either one. I never understood why as the obvious solution is to not build FPU test if !BUILD_FLAT. As arm64 has -Werror enabled, it blocks CI from passing.

Would it be OK to simply remove FPU test from the build if !BUILD_FLAT ?

**Comment by xiaoxiang781216** [comment]
> `./cxxtest_main.cxx:31:10: fatal error: fstream: No such file or directory` I think this is due to a missing C++ library, not sure why my build passes locally but CI fails. I think I'll just disable cxxtest.
> 
> ```
> Error: fpu.c:55:8: error: #warning "FPU test not built; Only available in the flat build (CONFIG_BUILD_FLAT)" [-Werror=cpp]
>    55 | #      warning "FPU test not built; Only available in the flat build (CONFIG_BUILD_FLAT)"
>       |        ^~~~~~~
> cc1: all warnings being treated as errors
> ```
> 
> This I'm not sure how to fix; ostest builds the FPU test for BUILD_KERNEL and BUILD_PROTECTED and then gives this warning about how the test does not support either one. I never understood why as the obvious solution is to not build FPU test if !BUILD_FLAT. As arm64 has -Werror enabled, it blocks CI from passing.
> 
> Would it be OK to simply remove FPU test from the build if !BUILD_FLAT ?
> 
> Or maybe just use this: https://github.com/apache/nuttx-apps/blame/c4d97cd8f8ad48ef5c68c75a271152cb681460d9/testing/ostest/Kconfig#L83-L85

Yes, I think the warning is unnecessary and boring.

File: arch/arm64/src/common/arm64_addrenv_perms.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  if (!l1vaddr)
+    {
+      /* Oops, there is no address environment to modify at all ? */
+
+      return -EINVAL;
+    }
+
+  /* Invalidate D-Cache so that we read from the physical memory */
+
+  up_invalidate_dcache(vstart, vend);
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_addrenv_perms.c:73]: This is wrong, we want to read the page tables from RAM, not the user physical memory.

File: arch/arm64/src/common/arm64_addrenv_perms.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+      entry |= setmask | PTE_BLOCK_DESC_AP_USER;
+
+      /* Restore the entry */
+
+      mmu_ln_restore(ptlevel, lnvaddr, vaddr, entry);
+    }
+
+  /* When all is set and done, flush the data cache */
+
+  up_flush_dcache(vstart, vend);
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_addrenv_perms.c:102]: Also wrong

File: arch/arm64/src/common/arm64_addrenv_utils.c [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+          /* Wipe the page and assign it */
+
+          arm64_pgwipe(paddr);
+          mmu_ln_setentry(ptlevel, ptprev, paddr, vaddr, flags);
+        }
+    }
+
+  /* Flush the data cache, so the changes are committed to memory */
+
+  up_flush_dcache(ptprev, sizeof(uintptr_t));
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_addrenv_utils.c:108]: Also wrong, what must be flushed is the entry in ptprev, not the first index

**Comment by pussuw** [comment]
The cache issues and build issues should now be fixed @acassis

**Review by acassis — APPROVED** [review]

**Pull request merged**